miraclelinux/meta-debian-extended
# Add valgrind package.

# Purpose of pull request

valgrind 

This pr adds valgrind package based on poky's meta/recipes-devtools/valgrind/valgrind_3.14.0.bb

# Test
## How to test

Prepare the following aarch64 program that exits without freeing memory.
mleaktest.c
```
#include <stdlib.h>

int main(void)
{
        char *buffer = (char *) malloc(1024);
        buffer[0] = 1;
        return 0;
}
```
```
$ aarch64-linux-gnu-gcc -o mleaktest mleaktest.c
```

Add valgrind and openssh (for file transfer), rebuild the image, and start qemu.
```
$ echo 'IMAGE_INSTALL_append = " valgrind openssh "' >> conf/local.conf
$ bitbake core-image-minimal
$ MACHINE=qemuarm64 runqemu nographic
```
Once booted, transfer from the host to qemu with the scp command.
```
$ scp mleaktest root@192.168.7.2:~/
```

Use the following command to check the operation.

### Memory leak detected
```
# valgrind --leak-check=full ./mleaktest
```
### Generating a callgraph
```
# valgrind --tool=callgrind ls
```
### Generating a heap profile
```
# valgrind --tool=massif ls
```

## Test result

**The validity of the output results has not been checked.**

### Memory leak detected
```
root@qemuarm64:~# valgrind --leak-check=full ./mleaktest
==2117== Memcheck, a memory error detector
==2117== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==2117== Using Valgrind-3.14.0 and LibVEX; rerun with -h for copyright info
==2117== Command: ./mleaktest
==2117==
==2117==
==2117== HEAP SUMMARY:
==2117==     in use at exit: 1,024 bytes in 1 blocks
==2117==   total heap usage: 1 allocs, 0 frees, 1,024 bytes allocated
==2117==
==2117== 1,024 bytes in 1 blocks are definitely lost in loss record 1 of 1
==2117==    at 0x4848198: malloc (vg_replace_malloc.c:299)
==2117==    by 0x4005A3: main (in /home/root/mleaktest)
==2117==
==2117== LEAK SUMMARY:
==2117==    definitely lost: 1,024 bytes in 1 blocks
==2117==    indirectly lost: 0 bytes in 0 blocks
==2117==      possibly lost: 0 bytes in 0 blocks
==2117==    still reachable: 0 bytes in 0 blocks
==2117==         suppressed: 0 bytes in 0 blocks
==2117==
==2117== For counts of detected and suppressed errors, rerun with: -v
==2117== ERROR SUMMARY: 1 errors from 1 contexts (suppressed: 0 from 0)
```
### Generating a callgraph
```
root@qemuarm64:~# valgrind --tool=callgrind ls
==2882== Callgrind, a call-graph generating cache profiler
==2882== Copyright (C) 2002-2017, and GNU GPL'd, by Josef Weidendorfer et al.
==2882== Using Valgrind-3.14.0 and LibVEX; rerun with -h for copyright info
==2882== Command: ls
==2882==
==2882== For interactive control, run 'callgrind_control -h'.
callgrind.out.2882  mleaktest
==2882==
==2882== Events    : Ir
==2882== Collected : 211049
==2882==
==2882== I   refs:      211,049
root@qemuarm64:~# vi callgrind.out.*
# callgrind format
version: 1
creator: callgrind-3.14.0
pid: 2882
cmd:  ls
part: 1


desc: I1 cache:
desc: D1 cache:
desc: LL cache:

desc: Timerange: Basic block 0 - 52870
desc: Trigger: Program termination

positions: line
events: Ir
summary: 211049


ob=(6) /bin/busybox.nosuid
fl=(68) ???
fn=(572) 0x00000000004730d0
... <snip> ...
```
### Generating a heap profile
```
root@qemuarm64:~# valgrind --tool=massif ls
==3268== Massif, a heap profiler
==3268== Copyright (C) 2003-2017, and GNU GPL'd, by Nicholas Nethercote
==3268== Using Valgrind-3.14.0 and LibVEX; rerun with -h for copyright info
==3268== Command: ls
==3268==
callgrind.out.2882  mleaktest
==3268==

root@qemuarm64:~# vi massif.out.*
desc: (none)
cmd: ls
time_unit: i
#-----------
snapshot=0
#-----------
time=0
mem_heap_B=0
mem_heap_extra_B=0
mem_stacks_B=0
heap_tree=empty
#-----------
snapshot=1
#-----------
time=148647
mem_heap_B=96
mem_heap_extra_B=8
mem_stacks_B=0
heap_tree=empty
#-----------
snapshot=2
#-----------
time=149733
... <snip> ...
```
